Add support for custom states

KonnorRogers opened the pull request:
Fixes #246 

I went with `@customState` instead of `@state` to disambiguate from `@state()` decorator from Lit.

thepassle commented on packages/analyzer/src/features/analyse-phase/creators/createClass.js:
```diff
@@ -24,7 +24,8 @@ export function createClass(node, moduleDoc, context) {
     events: [],
-    attributes: []
+    attributes: [],
+    customStates: [],
```
This is not standardized in the schema, right? If not we should add it there first

KonnorRogers replied:
Correct. In fact turns out @luke_warlow has a PR for it here:

https://github.com/webcomponents/custom-elements-manifest/pull/128

Looks like they went with:

`cssStates` so I'm guessing we should do the same. It also hasn't been merged or reviewed.

thepassle replied:
Lets go with whats suggested in the schema, and lets re-use that for the jsdoc tag

I left a comment on the PR on the schema and pinged Justin to ask him to also look at it

lukewarlow replied:
Iirc I went with cssStates because I want to use the manifest for defining built-in elements too. But I'm not super attached to the naming.

vospascal commented:
> Fixes #246
> 
> I went with `@customState` instead of `@state` to disambiguate from `@state()` decorator from Lit.

this for me feels correct but also not attached to a name as long as its specific enough
if your looking bit in to the future there going to be more exposed things other then parts and these customState/cssState
might be hard to keep these consistant like cross-root-aria proposal if the exportids is uses then you would get ids :)

KonnorRogers commented:
> should we also analyze for `this.internals.states.add("foo")`?
> 
> During class analysis we can see if `this.attachInternals()` is assigned to something, keep that something in memory, and then see if `<whatever>.states.add("foo")` is being called with a string (`"foo"`) on it?

Sounds similar to how events are "discovered", so I don't see why not

thepassle commented:
@KonnorRogers https://github.com/webcomponents/custom-elements-manifest/pull/128 was merged, can you update this PR? :)

KonnorRogers commented:
@thepassle updated to use `@cssstate` and `@cssState` pushing to `cssStates` array. 

I also updated the README to reflect the new supported tag.

I also added a new fixture directory for it.

thepassle commented:
Thanks for the PR! Published in 0.10.0

andrico1234 commented on docs/analyzer/getting-started.md:
```diff
@@ -331,18 +331,19 @@ For all supported syntax, please check the [fixtures](https://github.com/open-wc
+| `@part`,<br>`@csspart`        | Documents your custom elements Shadow Parts                                       |
+| `@slot`                       | Documents the Slots used in your components                                       |
+| `@cssprop`,<br>`@cssproperty` | Documents CSS Custom Properties for your component                                |
+| `@cssState`                   | Documents the custom CSS state of your custom element when using ElementInternals |
```
I'm a little confused, the code description for `@cssState` describes it as the "custom CSS state of your custom element when using ElementInternals".

After looking at the [API in MDN](https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals/states), there's no mention about CSS, the language, only CustomStateSet.

Is this part of the CSS spec and MDN just isn't clear about this, or is it just confusing that both abbreviations are `css`, if so, would it make sense to use a different abbreviation, or clarify this?

Let me know if I've misunderstood something :)

This is a cool change and I didn't know this ElementInternals feature existed, so this is more about me learning about new features than a criticism of the API design

andrico1234 replied:
Ignore me, I read the changes [here](https://github.com/webcomponents/custom-elements-manifest/pull/128) and it cleared things up :)

Still cool to hear about this feature! Awesome
